Feat: New homepage design

yangkb09 opened the pull request:
## description
in this PR, i implement the new homepage design based off @catherineymgui's [figma design](https://www.figma.com/design/eL7u2QrvLykWRLeUELiNNy/%5BGA%5D-Explore-Metrics---Master-File?node-id=1208-15616&node-type=instance&m=dev)

## old homepage vs new homepage comparison:
### old homepage (first time user)
<img width="1512" alt="Screenshot 2024-10-25 at 4 12 21 PM" src="https://github.com/user-attachments/assets/bd88bff4-a4e7-441b-803a-bfdca4449b1f">

### new homepage (first time user)
<img width="1512" alt="Screenshot 2024-10-25 at 4 10 47 PM" src="https://github.com/user-attachments/assets/cedc7e72-9c12-49ed-9f4d-5836691ef91f">

### old homepage (existing user)
![Screenshot 2024-10-25 at 4 33 25 PM (2)](https://github.com/user-attachments/assets/1d7873dd-e1f0-478f-9ab2-9dce9118069f)

### new homepage (existing user)
![Screenshot 2024-10-25 at 4 35 07 PM (2)](https://github.com/user-attachments/assets/59eb0b73-e3f8-48b3-9dfa-3a0a088a67b3)

## new features:

### new homepage (first time user)
- first time user will land directly on this page, only seeing the "billboard"
- format learn more and let’s start buttons
- hide recent metrics and bookmarks headers (and horizontal line)

### new homepage (recent explorations)
- grid layout
- new card UI
  - truncate singular label with long name
  - truncate multiple long labels after 3 lines
  - selecting a card takes you to the metric
- display only 3 recent metrics, can click "show more" button to show additional recent metrics

### bookmarks
- same as recent explorations:
  - grid layout
  - new card UI
- position delete button in line with data source line, but off to the right
- don't show section if no bookmarks
- collapse section by default
- expand section by clicking on arrow icon
- when expanded, arrow icon toggles up

### functionality
- select metric: recent metric card with "select metric" as the metric name is no longer created when users who only go to the select metric page (page showing lists of metrics) but do not click into a metric

### refactor
- refactored bookmarks and recent metrics into functional components

## future improvements:
- add unit tests
- all the discussion comments below
- new card UI: 
  - sometimes when multiple long labels get truncated, the labels end with a `|` => expected behavior is it should not end in a `|` but rather `-end-of-label...`
  - responsive card height to account for multiple labels spanning more than 3 lines OR implement hovering on labels to view the truncated labels 
- bookmarks: toggle icon to upward arrow when expanded

ref [#91815](https://github.com/grafana/grafana/issues/91815)

yangkb09 committed:
wip: homepage (new user)

yangkb09 committed:
fix: spacing between button and text; wip refactor: separating bookmarks and recent metrics from home page

yangkb09 committed:
feat: new user homepage; wip: need to clean up code

yangkb09 committed:
fix: change rocket icon to svg

yangkb09 committed:
wip feat: rendering recent metrics

yangkb09 committed:
chore: add comments to understand code, will need to delete / cleanup later / pare down into documentation comments

yangkb09 committed:
wip: new recent metric card design

yangkb09 committed:
wip: display recent metrics cards in rows of 3 (height still incorrect)

yangkb09 committed:
feat: apply conditional styling to remainder recent metrics exploration cards (any cards that are not a complete row of 3)

yangkb09 committed:
fix: render new recent metrics explorations without refresh

yangkb09 committed:
style: render recent metrics explorations in rows of 3 using grid instead of flex; fix: remove remainder card styling

yangkb09 committed:
fix: remove delete button from recent metrics exp cards

yangkb09 committed:
style: make background color for each card take up the entire card/grid space; make height of cards for each row the tallest card

yangkb09 committed:
chore: clean up code

yangkb09 committed:
fix: fix eslint errors

yangkb09 committed:
style: implement recent metrics card header styling

yangkb09 committed:
style: in recent metrics exp cards, format datasource line

yangkb09 committed:
fix: add initial value for _lastModified to fix eslint err

yangkb09 committed:
style: format date correctly; chore: clean up code; wip: get date to render properly on bottom left

yangkb09 committed:
style: make inner card height match outer card height; style: add date footer; style: wrap last metric name; style: wrap labels

yangkb09 committed:
style: adjust font for label name and label value

yangkb09 committed:
style: truncate singular label if value is greater than 35 characters

yangkb09 committed:
style: truncate singular long labels at 44 characters; style: truncate multiple labels at 3 lines; style: correct the border width and radius

yangkb09 committed:
style: make background border radius match the border

yangkb09 committed:
style: correct gap between rows and columns of cards; style: correct padding inside card

yangkb09 committed:
chore: clean up code

yangkb09 committed:
refactor: apply new card UI to DataTrailCard component

yangkb09 committed:
feat: add bookmarks (not formatted correctly), only render section if there are bookmarks, hook up delete functionality

yangkb09 committed:
style: add horizontal line above bookmarks header; style: add bookmarks header

yangkb09 committed:
style: add additional padding above bookmarks divider; chore: delete unused code

yangkb09 committed:
style: add carrot button to bookmarks; style: format heading font style

yangkb09 committed:
refactor: separate bookmarks into functional component; feat: make bookmarks section collapsed by default; feat: allow toggle to expand bookmarks section

yangkb09 committed:
style: position delete button for bookmarks in bottom right of card

yangkb09 committed:
fix: only render recent metrics and bookmarks headings if there are any

yangkb09 committed:
style: add show more button (not functional); style: fix padding around show more button

yangkb09 committed:
chore: delete unused code

yangkb09 committed:
fix: add back gap underneath bookmarks header

itsmylife commented:
BugBash: Button hover state needs some spacing 
![image](https://github.com/user-attachments/assets/0e09ed86-7064-4e1d-b9c4-55899a5f9d3e)

itsmylife commented:
BugBash: Really long metrics names are looking crazy on the card. I think they need some truncating too. 
![image](https://github.com/user-attachments/assets/e12bd45a-9e1a-4936-803b-03c012fa7c2b)

itsmylife commented:
BugBash: (some more findings)
- Pipe color is different than the design
- In the design it says "Last metric" but the implementation uses "Metric"
- Metric and labels should not have space in between

zhehao-grafana commented:
I assume the 3 additional recent metrics will get fixed by _(1) Show more button OR stop at 3 cards_
![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/cf4d5cec-1037-42e6-8737-f78c19da11f9)

NWRichmond commented:
I just found this Saga design system resource: https://grafana.com/developers/saga/templates/lists-of-objects/#tiled-cards. With that info in mind, I believe there's an opportunity to replace some of the CSS grid work with the `<Grid />` component (see [Saga](https://grafana.com/developers/saga/templates/grid-page) and [Storybook](https://developers.grafana.com/ui/latest/index.html?path=/docs/general-layout-grid--docs)). With `columns`, `minColumnWidth`, `gap`, and `alignItems` props, it sounds like it could do everything that's needed to achieve the design.

This might not be crucial to the initial implementation, but when the window is resized and becomes narrow, the right & left columns become cut off:

![exploremetrics homepage columns with narrow window](https://github.com/user-attachments/assets/0750dbfb-41eb-4d81-b589-ac9b327656c1)

Maybe the `<Grid />` component would help with this 🤔

NWRichmond commented on public/app/features/trails/DataTrailCard.tsx:
```diff
@@ -50,65 +59,168 @@ export function DataTrailCard(props: Props) {
-            </i>
+        </Card.Meta>
+        <div className={styles.datasource}>
+          <div className={styles.secondaryFont}>Datasource: </div>
```
Not sure if this matters, but the designs have "Data source:" instead of "Datasource:"
```suggestion
          <div className={styles.secondaryFont}>Data source: </div>
```

yangkb09 replied:
ah great catch!

yangkb09 committed:
feat: implement show more/less button for recent metrics

yangkb09 committed:
fix: do not show select metric card if user does not actually select a metric

yangkb09 committed:
chore: preliminary code clean up

yangkb09 committed:
Merge branch 'main' into kat/exm-homepage

yangkb09 commented:
i'm not sure why i have changes in `go.work.sum` and `.betterer.results`, this happened after i merged main into my branch to fix a few merge conflicts

yangkb09 committed:
chore: delete console.logs, comments

yangkb09 committed:
chore: clean up styling

yangkb09 committed:
Update public/app/features/trails/DataTrailCard.tsx

Co-authored-by: Nick Richmond <5732000+NWRichmond@users.noreply.github.com>

yangkb09 committed:
fix: add i18nKey to Trans tags

yangkb09 committed:
Merge branch 'main' into kat/exm-homepage

yangkb09 committed:
fix: attempt to remove go.work.sum changes that are unrelated to my PR

yangkb09 committed:
fix: add Trans tags

yangkb09 commented:
> BugBash: (some more findings)
> 
> * Pipe color is different than the design
> * In the design it says "Last metric" but the implementation uses "Metric"
> * Metric and labels should not have space in between

re: the last 2 bullets, @catherineymgui and i agreed on some changes during a meeting - what we agreed on is reflected in the current homepage so it's okay to disregard those 2 discrepancies between homepage and figma design :)

yangkb09 committed:
refactor: sepearate recent metrics into functional component; chore: delete unused code; fix: add Trans tags

yangkb09 commented:
i did notice a bug while testing - this behavior was happening prior to my changes too. basically:

click onto Metric A => use trail crumbs to go back to metric select view => select Metric B => click into homepage (via top nav bar) => only Metric B shows on recent metrics

do we want both Metric A and Metric B to show in the recent metrics section?

yangkb09 committed:
chore: generate translation json

yangkb09 committed:
trigger drone

yangkb09 committed:
trigger drone

itsmylife commented on public/app/features/trails/DataTrail.tsx:
```diff
@@ -137,8 +137,8 @@ export class DataTrail extends SceneObjectBase<DataTrailState> {
 
-    // Save the current trail as a recent if the browser closes or reloads
-    const saveRecentTrail = () => getTrailStore().setRecentTrail(this);
+    // Save the current trail (if it has a metric) as a recent if the browser closes or reloads
+    const saveRecentTrail = () => this.state.metric && getTrailStore().setRecentTrail(this);
```
It's quite a nice solution to prevent saving here in one place. But without a metric we also prevent saving a search with bunch of filters and time range applied. I don't think we want that. Please verify this behavior is the desired one. If not you can check the history step count and if it is just one that means it is a fresh selection and just don't save.

yangkb09 replied:
ah great catch, thank you! will address when pairing with brendan tomorrow

itsmylife commented on public/app/features/trails/DataTrailBookmarks.tsx:
```diff
@@ -0,0 +1,95 @@
+}
+export function DataTrailsBookmarks({ model, onDelete }: Props) {
+  const [toggleBookmark, setToggleBookmark] = useState(false);
+  const styles = useStyles2(getStyles);
+  return (
```
```suggestion
  const styles = useStyles2(getStyles);

  return (
```

itsmylife replied:
Just a new line between return statement.

itsmylife commented on public/app/features/trails/DataTrailCard.tsx:
```diff
@@ -50,68 +60,140 @@ export function DataTrailCard(props: Props) {
+      borderRight: `1px solid var(--border-Weak, rgba(204, 204, 220, 0.12))`,
+      borderLeft: `1px solid var(--border-Weak, rgba(204, 204, 220, 0.12))`,
+      borderBottom: 'none', // Remove the bottom border
+      // eslint-disable-next-line @grafana/no-border-radius-literal
+      borderRadius: '4px 4px 0 0', // Top-left and top-right corners are 4px, bottom-left and bottom-right are 0; cannot use theme.shape.radius.default because need bottom corners to be 0
```
Maybe there is a mistake in the UI design. @catherineymgui Could you please check the design. 
Having an eslint rule for borderRadius is weird enough and having it disabled is even weirder :D
But let's note this down to somewhere to check after merging the PR. Not a blocker. But it would be nice to understand why we have that rule in the first place and how we can play with it nicely.

catherineymgui replied:
Which part of the design is this? Not sure where there should be 4px border radius

yangkb09 replied:
in the new card UI, this is to make the top left and top right corners of the cards rounded, while keeping the bottom left and bottom right of the cards a right angle. i'll demonstrate what i mean in the weekly today!

NWRichmond replied:
Can this be accomplished with `theme.shape.radius[something]`?

yangkb09 replied:
basically, the way i have the card set up is that part A is one section and part B is a separate section. i did this to implement the different background colors between parts A and B. so part A needs to have it's top left and top right corners rounded to 4px, but the bottom left & right corners need to be a right angle (0px) since that's where part A meets part B

part B will have the top left and right corners as right angles, and the bottom left/right corners as rounded 4px
![462A9AC1-A0D0-457D-ACD2-46164FA25EB2](https://github.com/user-attachments/assets/4774271a-71e9-42b5-80a7-76fa937e3080)

if that doesn't make sense i can clarify in the daily tomorrow :)

catherineymgui replied:
Yeah this makes sense. I believe in my designs it's 2px corner radius instead of 4px tho (top left and right)?

For more context, the team updated all the corner radii across Grafana from 4px to 2px sometime ago.

itsmylife commented on public/app/features/trails/DataTrailsRecentMetrics.tsx:
```diff
@@ -0,0 +1,81 @@
+    setShowAll(!showAll);
+  };
+  return (
+    recentMetrics.length > 0 && (
```
If recentMetrics length is 0 then let's return null immediately.

itsmylife commented on public/app/features/trails/DataTrailsRecentMetrics.tsx:
```diff
@@ -0,0 +1,81 @@
+  const [showAll, setShowAll] = useState(false);
+  const handleToggleShow = () => {
+    setShowAll(!showAll);
+  };
+  return (
```
```suggestion
  };

  return (
```

itsmylife commented on public/app/features/trails/DataTrailBookmarks.tsx:
```diff
@@ -0,0 +1,95 @@
+
+interface Props extends SceneComponentProps<DataTrailsHome> {
+  onDelete: (index: number) => void;
+}
+export function DataTrailsBookmarks({ model, onDelete }: Props) {
```
```suggestion
}

export function DataTrailsBookmarks({ model, onDelete }: Props) {
```

itsmylife reviewed (commented):
In general it looks good and works nice. There are just some nitpicks. 
Great job 🚀

yangkb09 commented:
> I just found this Saga design system resource: https://grafana.com/developers/saga/templates/lists-of-objects/#tiled-cards. With that info in mind, I believe there's an opportunity to replace some of the CSS grid work with the `<Grid />` component (see [Saga](https://grafana.com/developers/saga/templates/grid-page) and [Storybook](https://developers.grafana.com/ui/latest/index.html?path=/docs/general-layout-grid--docs)). With `columns`, `minColumnWidth`, `gap`, and `alignItems` props, it sounds like it could do everything that's needed to achieve the design.
> 
> This might not be crucial to the initial implementation, but when the window is resized and becomes narrow, the right & left columns become cut off:
> 
> ![exploremetrics homepage columns with narrow window](https://private-user-images.githubusercontent.com/5732000/378519378-0750dbfb-41eb-4d81-b589-ac9b327656c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4ODY5MDcsIm5iZiI6MTcyOTg4NjYwNywicGF0aCI6Ii81NzMyMDAwLzM3ODUxOTM3OC0wNzUwZGJmYi00MWViLTRkODEtYjU4OS1hYzliMzI3NjU2YzEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjVUMjAwMzI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTIwYjMyMDg2OWI2NzliYmU0YTc3ZTY4Nzc5NTkxZTY5ODlhYWVlOTE2ODRlYjhlMDA0YzU0YzU0ZGI5NmFjZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2fIctFaURh8zlGRE6JNFzuIBLjWFc8xetp5mEGd6dAQ)
> 
> Maybe the `<Grid />` component would help with this 🤔

@NWRichmond this is very cool! noted - will look into if we decide to dedicate a bit more time to polishing things up :)

catherineymgui commented:
This is looking really good!

Feedback:

1. Currently light mode looks like this. Would check the tokens! I added a figma design for light mode too: https://www.figma.com/design/eL7u2QrvLykWRLeUELiNNy/%5BGA%5D-Explore-Metrics---Master-File?node-id=1522-8195&t=XFvfzGG6x4adz7as-1 

<img width="1375" alt="Screenshot 2024-10-25 at 13 30 52" src="https://github.com/user-attachments/assets/606faf8f-fd0a-4237-8222-f7ec9087d3b4">

2. The hover state on the arrow is overlapping with the text
<img width="283" alt="Screenshot 2024-10-25 at 13 32 24" src="https://github.com/user-attachments/assets/2ce852a3-99d1-48ba-8520-76240d0c7ab8">

3. I know you implemented the padding to spec, but I have made some mistakes in them. They should all be multiples of 4/8px.
So the gap should be 32 instead of 31
<img width="1228" alt="Screenshot 2024-10-25 at 13 46 08" src="https://github.com/user-attachments/assets/334322a6-6933-485c-959d-9bc39183261b">

A the bottom margin should be 24 instead of 20:
<img width="1078" alt="Screenshot 2024-10-25 at 13 45 18" src="https://github.com/user-attachments/assets/93c6bc0c-78bb-4b92-b2e5-e178575791dc">

4. I'm not sure if we can solve for this, but noticed that not only does the page jump left (I assume due to the scroll bar), but also a bit to the top when the bookmark section is expanded

https://github.com/user-attachments/assets/f70731fc-51f6-465d-943c-9490e7407192

5. I noticed explorations without a metric selected aren't saved in recents anymore - but they are also not showing if labels are added. See [this PR](https://github.com/grafana/grafana/issues/93073) for more info 

As a fyi I feel like the sections are a bit tight ever since we added the bookmarks (I know this is by design) but I will think about this a bit more

yangkb09 committed:
fix: add trans tag to date

yangkb09 committed:
chore: abbreviate descriptive key, regenerate json

yangkb09 committed:
Update public/app/features/trails/DataTrailBookmarks.tsx

Co-authored-by: ismail simsek <ismailsimsek09@gmail.com>

yangkb09 committed:
Update public/app/features/trails/DataTrailsRecentMetrics.tsx

Co-authored-by: ismail simsek <ismailsimsek09@gmail.com>

yangkb09 committed:
Update public/app/features/trails/DataTrailBookmarks.tsx

Co-authored-by: ismail simsek <ismailsimsek09@gmail.com>

yangkb09 committed:
fix: revert trans tag on date created to fix formatting

yangkb09 committed:
chore: return null immediately if no recent metrics

yangkb09 committed:
style: add margin between  bookmarks header and carrot toggle button

yangkb09 committed:
style: adjust margin to 8px between bookmarks header and carrot toggle button

yangkb09 committed:
style: make margins multiples of 4

yangkb09 commented:
> This is looking really good!
> 
> Feedback:
> 
> 1. Currently light mode looks like this. Would check the tokens! I added a figma design for light mode too: https://www.figma.com/design/eL7u2QrvLykWRLeUELiNNy/%5BGA%5D-Explore-Metrics---Master-File?node-id=1522-8195&t=XFvfzGG6x4adz7as-1
> 
> <img alt="Screenshot 2024-10-25 at 13 30 52" width="1375" src="https://private-user-images.githubusercontent.com/5264102/380319501-606faf8f-fd0a-4237-8222-f7ec9087d3b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMjc4MjIsIm5iZiI6MTczMDEyNzUyMiwicGF0aCI6Ii81MjY0MTAyLzM4MDMxOTUwMS02MDZmYWY4Zi1mZDBhLTQyMzctODIyMi1mN2VjOTA4N2QzYjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjhUMTQ1ODQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTkzYmM1OGY3NmUyZWRhMTI3M2ZiYjM2ODY5ZmJjNDNjYWUwZjEwYzcxMjZhN2UzZjRhOGI5ZmJiNjE3Y2QxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.iN08Pxk8emiWrrRLxTIc74XwzWFEjNLU2Mf04H8J-4Y">
> 2. The hover state on the arrow is overlapping with the text
> 
> <img alt="Screenshot 2024-10-25 at 13 32 24" width="283" src="https://private-user-images.githubusercontent.com/5264102/380320555-2ce852a3-99d1-48ba-8520-76240d0c7ab8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMjc4MjIsIm5iZiI6MTczMDEyNzUyMiwicGF0aCI6Ii81MjY0MTAyLzM4MDMyMDU1NS0yY2U4NTJhMy05OWQxLTQ4YmEtODUyMC03NjI0MGQwYzdhYjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjhUMTQ1ODQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWU5OGE4MzgxOWQ3NWQ5OWI3YzE0ZWRhNGQxNDZkY2IwMDYxMGRlMjAyNTFiZWIzN2QxODc3ZmRkYjJlY2Y0YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.eh8OtgKWSnWlrF55hgf-4UBoFYBTZurEQNWKMGHMv1A">
> 3. I know you implemented the padding to spec, but I have made some mistakes in them. They should all be multiples of 4/8px.
>    So the gap should be 32 instead of 31
> 
> <img alt="Screenshot 2024-10-25 at 13 46 08" width="1228" src="https://private-user-images.githubusercontent.com/5264102/380321507-334322a6-6933-485c-959d-9bc39183261b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMjc4MjIsIm5iZiI6MTczMDEyNzUyMiwicGF0aCI6Ii81MjY0MTAyLzM4MDMyMTUwNy0zMzQzMjJhNi02OTMzLTQ4NWMtOTU5ZC05YmMzOTE4MzI2MWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjhUMTQ1ODQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjFjMWE2NmE2MjJkOWRjZjYxNzFiMTkyYjkwMTE4MzgzZmU3NWJlNGViOGI1OTY5MDI5NDFhZjVmZDA4YTI5OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qxXM5tcWUWyoswVLEz5WVKNHmZiyoAt9quMxAlWEqfw">
> A the bottom margin should be 24 instead of 20: <img alt="Screenshot 2024-10-25 at 13 45 18" width="1078" src="https://private-user-images.githubusercontent.com/5264102/380321395-93c6bc0c-78bb-4b92-b2e5-e178575791dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMjc4MjIsIm5iZiI6MTczMDEyNzUyMiwicGF0aCI6Ii81MjY0MTAyLzM4MDMyMTM5NS05M2M2YmMwYy03OGJiLTRiOTItYjJlNS1lMTc4NTc1NzkxZGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjhUMTQ1ODQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzY4NmExYWFjMmRmNzQwNjU4OWNkOWJhOWRiNTMwZDc5NDIwNjEwMGVkMmZlNGMzNDMyNmE5ZTUwZGZiZDE2NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.SpRa0N7HNgnxGkV7mVrBDesuh1HAN6JGPalFqgwV7nw">
> 
> 4. I'm not sure if we can solve for this, but noticed that not only does the page jump left (I assume due to the scroll bar), but also a bit to the top when the bookmark section is expanded
> 
>  Screen.Recording.2024-10-25.at.13.51.01.mov 
> 5. I noticed explorations without a metric selected aren't saved in recents anymore - but they are also not showing if labels are added. See [this PR](https://github.com/grafana/grafana/issues/93073) for more info
> 
> As a fyi I feel like the sections are a bit tight ever since we added the bookmarks (I know this is by design) but I will think about this a bit more

@catherineymgui so far, i implemented:

- [ x ] #2
- [ x ] #3

bohandley commented on public/app/features/trails/DataTrail.tsx:
```diff
@@ -137,8 +137,8 @@ export class DataTrail extends SceneObjectBase<DataTrailState> {
-    // Save the current trail as a recent if the browser closes or reloads
-    const saveRecentTrail = () => getTrailStore().setRecentTrail(this);
+    // Save the current trail (if it has a metric) as a recent if the browser closes or reloads
+    const saveRecentTrail = () => this.state.metric && getTrailStore().setRecentTrail(this);
```
Check for those filters with this special Scenes variable lookup 👀 

```suggestion
    const saveRecentTrail = () => {
      const filtersVariable = sceneGraph.lookupVariable(VAR_FILTERS, this);
      const hasFilters = filtersVariable instanceof AdHocFiltersVariable && filtersVariable.state.filters.length > 0;

      return ((this.state.metric || hasFilters) 
      && getTrailStore().setRecentTrail(this))};```

bohandley replied:
The variable lookup is interesting 🤔 You might think the filters are stored in the state but this `VAR_FILTERS` variable is one way that Scenes can access stored data across all the Scenes components.

bohandley commented on public/app/features/trails/DataTrailCard.tsx:
```diff
@@ -17,6 +18,15 @@ export type Props = {
   onDelete?: () => void;
 };
 
+// Helper function to truncate the value for a single key:value pair
+const truncateValue = (key: string, value: string, maxLength: number) => {
+  const combinedLength = key.length + 2 + value.length; // 2 for ": "
+  if (combinedLength > maxLength) {
+    return value.substring(0, maxLength - key.length - 5) + '...'; // 5 for ": " and "..."
+  }
+  return value;
+};
+
```
Woohoo! Amazing 🤘🏻 Love this truncation

bohandley commented:
For the light mode issues, we can look at this file to replace any hardcoded styles and use the `theme.<thing>.<you>.<want>.<styled>`

https://github.com/grafana/grafana/blob/kat/exm-homepage/contribute/style-guides/themes.md#background-colors

bohandley reviewed (commented):
Great work @yangkb09! We can look at the things I commented on tomorrow. I found nothing that hadn't already been bug bashed. Great work @grafana/observability-metrics on bug bashing!!!

We'll get this all nice and sorted tomorrow. Well done, Kat!!! 👏🏻 👏🏻 👏🏻

yangkb09 committed:
Update public/app/features/trails/DataTrailBookmarks.tsx

Co-authored-by: Brendan O'Handley <brendan.ohandley@grafana.com>

yangkb09 committed:
style: fix light mode styles; style: fix border radius

yangkb09 committed:
fix: save select metric view as recent metric card if labels are applied

bohandley reviewed (approved):
Paired with @yangkb09 and fixed the light mode issue and the saving trail with filters. Happy to wait for Catherine and Nick's feedback but ✅ from me!

LGTM! 🚀

NWRichmond commented on public/app/features/trails/DataTrailBookmarks.tsx:
```diff
@@ -0,0 +1,97 @@
+  const [toggleBookmark, setToggleBookmark] = useState(false);
+  const styles = useStyles2(getStyles);
+
+  return (
+    <>
+      {getTrailStore().bookmarks.length > 0 && (
+        <>
+          <div className={styles.horizontalLine} />
+          <div className={css(styles.gap20, styles.bookmarkHeader, styles.bottomGap24)}>
+            <div className={styles.header} style={{ marginRight: '8px' }}>
+              <Trans i18nKey="trails.bookmarks.or-view-bookmarks">Or view bookmarks</Trans>
+            </div>
+            <IconButton
+              name={toggleBookmark ? 'angle-up' : 'angle-down'}
+              size="xxxl"
+              aria-label="bookmarkCarrot"
+              variant="secondary"
+              onClick={() => setToggleBookmark(!toggleBookmark)}
+            />
+          </div>
+          {toggleBookmark && (
+            <div className={styles.trailList}>
+              {getTrailStore().bookmarks.map((bookmark, index) => {
+                return (
+                  <DataTrailCard
+                    key={getBookmarkKey(bookmark)}
+                    bookmark={bookmark}
+                    onSelect={() => model.onSelectBookmark(index)}
+                    onDelete={() => onDelete(index)}
+                  />
+                );
+              })}
+            </div>
+          )}
+        </>
+      )}
+    </>
+  );
```
We can simplify the JSX with an early return:

```suggestion
  if (getTrailStore().bookmarks.length === 0) {
    return null;
  }

  return (
    <>
      <div className={styles.horizontalLine} />
      <div className={css(styles.gap20, styles.bookmarkHeader, styles.bottomGap24)}>
        <div className={styles.header} style={{ marginRight: '8px' }}>
          <Trans i18nKey="trails.bookmarks.or-view-bookmarks">Or view bookmarks</Trans>
        </div>
        <IconButton
          name={toggleBookmark ? 'angle-up' : 'angle-down'}
          size="xxxl"
          aria-label="bookmarkCarrot"
          variant="secondary"
          onClick={() => setToggleBookmark(!toggleBookmark)}
        />
      </div>
      {toggleBookmark && (
        <div className={styles.trailList}>
          {getTrailStore().bookmarks.map((bookmark, index) => {
            return (
              <DataTrailCard
                key={getBookmarkKey(bookmark)}
                bookmark={bookmark}
                onSelect={() => model.onSelectBookmark(index)}
                onDelete={() => onDelete(index)}
              />
            );
          })}
        </div>
      )}
    </>
  );
```

NWRichmond commented on public/app/features/trails/TrailStore/TrailStore.ts:
Are we using `lastModified` anywhere? If not, perhaps this change could be held back until we're using it.

yangkb09 replied:
i tried undoing the `lastModified` changes but then the recent metrics show more functionality stopped working, so for now we need it!

NWRichmond reviewed (commented):
This is coming along nicely! I'll take a look at light mode fixes shortly. In the meantime, I have some non-blocking suggestions & nits.

In addition to some feedback below on specific files, here are some thoughts about some overarching themes:
- Consider leveraging elements like [article](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article), [section](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section), etc. in place of a surrounding fragment or div.
- There are some styles (like `fontStyle: 'normal'` and `fontFamily: 'Inter'`) that don't appear to be changing anything. Removing these will reduce the size of the JS associated with our styles.
- There are some `div`s with styles that contain comments like `/* H4 */`, and I'm curious if these could actually be `<Text variant="h4">`s instead 🤔

yangkb09 committed:
Update public/app/features/trails/DataTrailCard.tsx

Co-authored-by: Nick Richmond <5732000+NWRichmond@users.noreply.github.com>

yangkb09 committed:
refactor: move rocket svgs into assets folder

yangkb09 commented on public/app/features/trails/DataTrailsHome.tsx:
i moved them to an assets folder!

yangkb09 committed:
chore: add back accidentally deleted console log

yangkb09 committed:
Update public/app/features/trails/DataTrail.tsx

Co-authored-by: Nick Richmond <5732000+NWRichmond@users.noreply.github.com>

yangkb09 committed:
Update public/app/features/trails/DataTrailBookmarks.tsx

Co-authored-by: Nick Richmond <5732000+NWRichmond@users.noreply.github.com>

yangkb09 committed:
chore: revert lastModified related changes since behavior appears to remain the same

NWRichmond reviewed (approved):
Light mode is looking beautiful!

![demo light mode explore metrics homepage redesign](https://github.com/user-attachments/assets/4013c9ca-cbd2-4be9-ae17-70a5e881e598)

Nicely done, @yangkb09. Everything appears to be working as expected. What a nice glow-up for Explore Metrics!

zhehao-grafana commented:
> Light mode is looking beautiful!
> 
> ![demo light mode explore metrics homepage redesign](https://private-user-images.githubusercontent.com/5732000/381305881-4013c9ca-cbd2-4be9-ae17-70a5e881e598.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyMzY1NDQsIm5iZiI6MTczMDIzNjI0NCwicGF0aCI6Ii81NzMyMDAwLzM4MTMwNTg4MS00MDEzYzljYS1jYmQyLTRiZTktYWUxNy03MGE1ZTg4MWU1OTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjlUMjExMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWYzMmNlOGMxYmJiZDlhOTMyNmZhYmY2OWU2Mzc4Nzc5MzIyMTQyNGE0YjU5ZDE4ODM0YjI0ZWMzNDFkNWYyNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hY2Lk37NNb8dnZnPO2iZJyHeXo9ElEgfYAu3uXCG_j4)
> 
> Nicely done, @yangkb09. Everything appears to be working as expected. What a nice glow-up for Explore Metrics!

It is truly beautiful and thanks for the great work from Kat and Brendan!

catherineymgui commented:
This is looking good based on my feedback!

Few things:
1. In this [issue](https://github.com/grafana/grafana/issues/93073) - we also mentioned changing the text of 'Select metric' to 'All metrics', for the saved recent explorations on the 'All metrics' page. if it's a quick thing, I think we could update this. Or we can leave it to another PR
2. Page jumping left - this seems to not show up in Chrome, only Arc Browser. But given that not many people use it, it's okay to ignore this
3. Based on our slack convo, we talked about a few things to improve the design of the page (since the recent metric cards are quite big, I'm finding that it's taking away from the 'Let's start') button. The proposal is to increase the size of the 'Let's start' section / card, and decrease the number of lines in the recent metric and bookmark cards. We can discuss if this should be in this PR, or another one. I will put this in slack too
<img width="1310" alt="Screenshot 2024-10-29 at 14 23 34" src="https://github.com/user-attachments/assets/c7ea936e-f824-40e8-824a-2d683ed81b6d">

yangkb09 committed:
fix: add back lastModified changes because they make the recent metrics show more functionality work
